samza-azure/src/test/java/org/apache/samza/system/azureblob/avro/TestAzureBlobOutputStream.java

```diff
@@ -0,0 +1,347 @@
+public class TestAzureBlobOutputStream {
+  private ThreadPoolExecutor threadPool;
+  private ByteArrayOutputStream mockByteArrayOutputStream;
+  private static final int THRESHOLD = 100;
+  private BlockBlobAsyncClient mockBlobAsyncClient;
+  private AzureBlobOutputStream azureBlobOutputStream;
+  private static String randomString = "roZzozzLiR7GCEjcB0UsRUNgBAip8cSLGXQSo3RQvbIDoxOaaRs4hrec2s5rMPWgTPRY4UnE959worEtyhRjwUFnRnVuNFZ554yuPQCbI69qFkQX7MmrB4blmpSnFeGjWKjFjIRLFNVSsQBYMkr5jT4T83uVtuGumsjACVrpcilihdd194H8Y71rQcrXZoTQtw5OvmPicbwptawpHoRNzHihyaDVYgAs0dQbvVEu1gitKpamzYdMLFtc5h8PFZSVEB";
+  private static byte[] bytes = randomString.substring(0, THRESHOLD).getBytes();
+  private static byte[] compressedBytes = randomString.substring(0, THRESHOLD / 2).getBytes();
+  private AzureBlobWriterMetrics mockMetrics;
+  private Compression mockCompression;
+
+  @Before
+  public void setup() throws Exception {
+    threadPool = new ThreadPoolExecutor(1, 1, 60,
+        TimeUnit.SECONDS, new LinkedBlockingDeque<Runnable>());
+
+
+    mockByteArrayOutputStream = spy(new ByteArrayOutputStream(THRESHOLD));
+
+    mockBlobAsyncClient = PowerMockito.mock(BlockBlobAsyncClient.class);
+    when(mockBlobAsyncClient.stageBlock(anyString(), any(), anyLong())).thenReturn(
+        Mono.just(new SimpleResponse(null, 200, null, null)).flatMap(FluxUtil::toMono));
+    when(mockBlobAsyncClient.commitBlockListWithResponse(any(), any(), any(), any(), any())).thenReturn(
+        Mono.just(new SimpleResponse(null, 200, null, null)));
+
+    when(mockBlobAsyncClient.getBlobUrl()).thenReturn("https://samza.blob.core.windows.net/fake-blob-url");
+
+    mockMetrics = mock(AzureBlobWriterMetrics.class);
+
+    mockCompression = mock(Compression.class);
+    doReturn(compressedBytes).when(mockCompression).compress(bytes);
+
+    azureBlobOutputStream = spy(new AzureBlobOutputStream(mockBlobAsyncClient, threadPool, mockMetrics,
+        60000, THRESHOLD, mockByteArrayOutputStream, mockCompression));
+  }
+
+  @Test
+  public void testWrite() {
+    byte[] b = new byte[THRESHOLD - 10];
+    azureBlobOutputStream.write(b, 0, THRESHOLD - 10);
+    verify(mockBlobAsyncClient, never()).stageBlock(any(), any(), anyLong()); // since size of byte[] written is less than threshold
+    verify(mockMetrics).updateWriteByteMetrics(THRESHOLD - 10);
+    verify(mockMetrics, never()).updateAzureUploadMetrics();
+  }
+
+  @Test
+  public void testWriteLargerThanThreshold() {
+    byte[] largeRecord = randomString.substring(0, 2 * THRESHOLD).getBytes();
+    byte[] largeRecordFirstHalf = randomString.substring(0, THRESHOLD).getBytes();
+    byte[] largeRecordSecondHalf = randomString.substring(THRESHOLD, 2 * THRESHOLD).getBytes();
+
+    byte[] compressB = randomString.substring(0, THRESHOLD / 2).getBytes();
+
+    doReturn(compressB).when(mockCompression).compress(largeRecordFirstHalf);
+    doReturn(compressB).when(mockCompression).compress(largeRecordSecondHalf);
```

Comment:
Please consider returning different "compressed bytes" for the first half and second half. It will make the test more complete to ensure that the correct bytes are used for each half.